rfcs/0003-gene-sets/api.md

```diff
@@ -0,0 +1,400 @@
+
+## Problem Statement | Background
+
+Gene sets are the expected output of differential expression analysis in cellxgene. Two (non-overlaping) sets of cells
+are selected and the difference in the expression level for each gene is computed by comparing the average
+expression level for each set of cells. Here I will define the structure of the API to ensure communication between the
+client (where the user is working directly with the data or uploading their own gene sets) and the backend (where data
+is stored to allow it to persist between sessions) is consistent and clear.
+
+## Product Requirements
+
+### Users should be able to export differential expression results
+
+Currently cellxgene users can choose two sets of cells and compute/display the log fold and adjusted p-value of the top ten differentially expressed genes. Additionally we display a histogram of the gene expression levels of the selected cells for each gene in the right side bar.
+
+- [US1] A user should be able to export a gene set, if the gene set was created by differential expression, the pvalue, log fold change, and category label fields will be populated in addition to the gene name. This will be available in a csv file (see data formats for more detail).
+
+### Users should be able to upload their own gene sets and see them in the right side bar with the option to apply them within current cellxgene functionalities, if the same gene ontology was used when generating the current dataset and the gene set uploaded by the user (ie the gene names match)
+
+- [US2] A user should be able to upload a gene set via a csv file (see data formats for more detail) for a dataset.
+- [US3] A user should be able to upload a gene set they previously downloaded from cellxgene
+- [US4] When a user uploads a gene set, that list of genes should appear in the right side bar.
+- [US5] If the user was logged in when they uploaded a gene set that gene set should persist if they logout and back in or reload the page
+- [US6] If the gene names match those used in the dataset the user should be able to use general cellxgene functionality
+  (color by genes, calculate differential expression levels, etc.) on those genes.
+- [US7] If a gene in the user uploaded gene set does not exist in the dataset the user will see an error message specifying the gene that wasnt in the dataset, but the remaining genes will be displayed.
+
+### If there is a standard format for differential expression output, we should follow that. If not, the format should be as straightforward as possible while serving these above cases
+
+### Nonfunctional requirements
+
+- Users should be able to export precomputed differential expression results in real time
+- Users should be able to upload gene sets and have them immediately appear in the right side bar
```

Comment:
Would delete this -- it's a functional requirement that's already written above.